blockindex/sgd_indexer.go

```diff
@@ -7,50 +7,448 @@ package blockindex
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "beacon",
+						"type": "address"
+					}
+				],
+				"name": "BeaconUpgraded",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					}
+				],
+				"name": "ContractApproved",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					},
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "recipient",
+						"type": "address"
+					}
+				],
+				"name": "ContractRegistered",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					}
+				],
+				"name": "ContractRemoved",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "address",
+						"name": "contractAddress",
+						"type": "address"
+					}
+				],
+				"name": "ContractDisapproved",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": false,
+						"internalType": "uint8",
+						"name": "version",
+						"type": "uint8"
+					}
+				],
+				"name": "Initialized",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "previousOwner",
+						"type": "address"
+					},
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "newOwner",
+						"type": "address"
+					}
+				],
+				"name": "OwnershipTransferred",
+				"type": "event"
+			},
+			{
+				"anonymous": false,
+				"inputs": [
+					{
+						"indexed": true,
+						"internalType": "address",
+						"name": "implementation",
+						"type": "address"
+					}
+				],
+				"name": "Upgraded",
+				"type": "event"
+			}
+]`
+)
+
+func init() {
+	var err error
+	_sgdABI, err = abi.JSON(strings.NewReader(_sgdContractInterfaceABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+const (
+	_sgdBucket     = "sg"
+	_sgdToHeightNS = "hh"
+	//TODO (millken): currently we fix the percentage to 30%, we can make it configurable in the future
+	_sgdPercentage = uint64(30)
+)
+
+var _sgdCurrentHeight = []byte("currentHeight")
+
 type (
 	// SGDRegistry is the interface for Sharing of Gas-fee with DApps
 	SGDRegistry interface {
 		blockdao.BlockIndexer
 		// CheckContract returns the contract's eligibility for SGD and percentage
 		CheckContract(context.Context, string) (address.Address, uint64, bool, error)
+		// FetchContracts returns all contracts that are eligible for SGD
+		FetchContracts(context.Context) ([]*indexpb.SGDIndex, error)
```

Comment:
should not use `*indexpb.SGDIndex`, which is storage format

Reply:
define a `struct sgdIndex`, and convert it to/from `*indexpb.SGDIndex`

Reply:
fixed, defined a  `SGDIndex`